### PR TITLE
Implement LWG-3590: `split_view::base() const &` is overconstrained

### LINKED ISSUE #2400: LWG-3590 `split_view::base() const &` is overconstrained
LWG-3590 `split_view::base() const &` is overconstrained

This was voted into the Working Paper in October 2021, see #2236.

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2400

### CODE REVIEW

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 1]
```diff
@@ -328,6 +329,23 @@ constexpr bool test_devcom_1559808() {
...
+};
+
+// COMPILE-ONLY
+void testLWG3590(const CopyConstructibleRange& r) {
+    auto split_view = views::split(r, 0);
+
+    STATIC_ASSERT(copy_constructible<CopyConstructibleRange>);
+    STATIC_ASSERT(!copyable<CopyConstructibleRange>);
+
+    [[maybe_unused]] auto base = split_view.base();
```
Comment (miscco) [inline tests/std/tests/P0896R4_views_split/test.cpp:346]: Should we test equality with the original range?

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 2]
```diff
@@ -328,6 +329,23 @@ constexpr bool test_devcom_1559808() {
...
 
+struct CopyConstructibleRange {
+    CopyConstructibleRange(const CopyConstructibleRange&) = default;
+    CopyConstructibleRange& operator=(const CopyConstructibleRange&) = delete;
+    int* begin() const;
+    int* end() const;
+};
+
+// COMPILE-ONLY
+void testLWG3590(const CopyConstructibleRange& r) {
```
Comment (miscco) [inline tests/std/tests/P0896R4_views_split/test.cpp:340]: I would sleep better if we would actually call it

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 3]
```diff
@@ -328,6 +329,23 @@ constexpr bool test_devcom_1559808() {
     return true;
 }
 
+struct CopyConstructibleRange {
+    CopyConstructibleRange(const CopyConstructibleRange&) = default;
+    CopyConstructibleRange& operator=(const CopyConstructibleRange&) = delete;
+    int* begin() const;
```
Comment (miscco) [inline tests/std/tests/P0896R4_views_split/test.cpp:335]: Can we just add a vector<int> so that we can check that it is actually copied

**Review by miscco — APPROVED** [review]
Mostly nitpicks

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 4]
```diff
@@ -328,6 +329,32 @@ constexpr bool test_devcom_1559808() {
     return true;
 }
 
+struct CopyConstructibleRange {
+    constexpr CopyConstructibleRange(vector<int> v) : inner(move(v)) {}
+    CopyConstructibleRange(const CopyConstructibleRange&) = default;
+    CopyConstructibleRange& operator=(const CopyConstructibleRange&) = delete;
+    constexpr const int* begin() const {
+        return inner.empty() ? nullptr : &inner[0];
```
Comment (miscco) [inline tests/std/tests/P0896R4_views_split/test.cpp:337]:
```suggestion
        return inner.data();
```

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 5]
```diff
@@ -328,6 +329,32 @@ constexpr bool test_devcom_1559808() {
...
 
+struct CopyConstructibleRange {
+    constexpr CopyConstructibleRange(vector<int> v) : inner(move(v)) {}
+    CopyConstructibleRange(const CopyConstructibleRange&) = default;
+    CopyConstructibleRange& operator=(const CopyConstructibleRange&) = delete;
+    constexpr const int* begin() const {
+        return inner.empty() ? nullptr : &inner[0];
+    }
+    constexpr const int* end() const {
+        return inner.empty() ? nullptr : (&inner[0] + inner.size());
```
Comment (miscco) [inline tests/std/tests/P0896R4_views_split/test.cpp:340]:
```suggestion
        return inner.data() + inner.size();
```

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 6]
```diff
@@ -328,6 +329,32 @@ constexpr bool test_devcom_1559808() {
...
+    constexpr const int* end() const {
+        return inner.data() + inner.size();
+    }
+
+    vector<int> inner;
+};
+
+constexpr bool test_LWG_3590() {
+    CopyConstructibleRange r{{1, 2, 3, 4}};
+    const auto split_view = views::split(r, 0);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_split/test.cpp:348]: No change requested: I observe that `split_view` and `ref_view` below are named identically to their types - however, they aren't technically shadowing as this test isn't `using namespace std::ranges;` so it's fine.

**Review by StephanTLavavej — APPROVED** [review]

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 7]
```diff
@@ -328,6 +329,32 @@ constexpr bool test_devcom_1559808() {
...
+
+    STATIC_ASSERT(copy_constructible<CopyConstructibleRange>);
+    STATIC_ASSERT(!copyable<CopyConstructibleRange>);
+
+    const auto ref_view = split_view.base();
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_views_split/test.cpp]: `CopyConstructibleRange` might not be `copyable`, but `ref_view<CopyConstructibleRange>` is - this isn't actually testing the fix.

**Review by CaseyCarter — APPROVED** [review]
I've pushed a fix for my comments.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this LWG issue resolution! :rocket: :tada: :smiley_cat:

**Pull request merged**